ramips: fix WiFi MAC addresses for D-Link DIR-810L

adschm opened the pull request:
So far, WiFi MAC addresses for this device have been set up from
caldata. However, this returns values which do not look like MAC
addresses. They also do not match stock firmware:

wlan0 (5.0): 00:11:22:00:17:D0 from 0x8004
wlan1 (2.4): 00:11:22:00:17:CD from 0x4 (and 0x2e)

It looks like the only valid MAC address on this device is at 0x28.

So, this patch changes setup to calculate addresses based on the
value at 0x28:

lan: *:0A (flash, label)
wan: *:0B (flash + 1)
wifi2: *:0A (flash)
wifi5: *:0C (flash + 2)

Thanks to Roger Pueyo Centelles <roger.pueyo@guifi.net> for
investigating this on his devices.

rogerpueyo commented:
Hi,

The 2.4 GHz radio now gets the correct *:0A MAC.

The 5 GHz radio part looks good to me but insists on using the one at 0x8004 (00:11:22:00:17:D0). Could it be an mt7610e-specific issue, maybe related with the driver?

adrianschmutzler commented:
> The 5 GHz radio part looks good to me but insists on using the one at 0x8004 (00:11:22:00:17:D0). Could it be an mt7610e-specific issue, maybe related with the driver?

Might be, supposed that this (with other hardware) is working:
https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/dts/mt7628an_tplink_archer-c20-v4.dts#L97-L103

adrianschmutzler commented:
This one states it's MT7610EN:
https://github.com/openwrt/openwrt/pull/1550

Despite, there is TP-Link Archer C20 v1:
https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/dts/mt7620a_tplink_archer-c20-v1.dts#L191-L197

adrianschmutzler commented:
@rogerpueyo Just for fun try to put the mac-address commands in front of mediatek,eeprom and put the increment in bracket, i.e.:

&pcie0 {
	wifi@0,0 {
		reg = <0x0000 0 0 0 0>;
		mtd-mac-address = <&factory 0x28>;
		mtd-mac-address-increment = <(2)>;
		mediatek,mtd-eeprom = <&factory 0x8000>;
		ieee80211-freq-limit = <5000000 6000000>;
	};

Don't believe in it, but just to make sure ...

adrianschmutzler commented:
.... and have a look whether dmesg tells you anything ....

adrianschmutzler commented:
blocktrron suggested this might help:
https://github.com/openwrt/mt76/pull/209

@rogerpueyo Can you test this by modifying mt76x0/eeprom.c locally as in the referred patch?

Otherwise I will do a patch-in-patch if I have some additional time.

adrianschmutzler commented:
I added a patch for that, which seems to build at least. Maybe just try that ...

rogerpueyo commented:
> I added a patch for that, which seems to build at least. Maybe just try that ...

It builds, and it works:
```
root@OpenWrt:~# iwinfo 
wlan0     ESSID: unknown
          Access Point: C8:D3:A3:76:CC:0C
          Mode: Client  Channel: unknown (unknown)
          Tx-Power: 0 dBm  Link Quality: unknown/70
          Signal: unknown  Noise: unknown
          Bit Rate: unknown
          Encryption: unknown
          Type: nl80211  HW Mode(s): 802.11nac
          Hardware: 14C3:7650 14C3:7650 [MediaTek MT7610E]
          TX power offset: none
          Frequency offset: none
          Supports VAPs: yes  PHY name: phy0

wlan1     ESSID: unknown
          Access Point: C8:D3:A3:76:CC:0A
          Mode: Client  Channel: unknown (unknown)
          Tx-Power: 0 dBm  Link Quality: unknown/70
          Signal: unknown  Noise: unknown
          Bit Rate: unknown
          Encryption: unknown
          Type: nl80211  HW Mode(s): 802.11bgn
          Hardware: 14C3:7620 14C3:000C [MediaTek MT7620]
          TX power offset: none
          Frequency offset: none
          Supports VAPs: yes  PHY name: phy1
```
Amazing! :+1: Thank you very much, @adrianschmutzler!

> Just for fun try to put the mac-address commands in front of mediatek,eeprom and

Nope, that didn't work :smile:

adrianschmutzler commented:
Okay, I've made the second patch a little nicer.

@blocktrron @nbd168 @CHKDSK88 

Please have a look.

rogerpueyo commented:
@adrianschmutzler,

Since the 5 GHz radio was not supported until recently, I hadn't been using the router too much with OpenWrt. Now, I've also realized that the 2.4 GHz radio had very bad signal figures (e.g. -89 dBm for an AP placed 5 m away). The &wmac entry was missing this two extra lines:
```
&wmac {
	ralink,mtd-eeprom = <&factory 0x0>;
	mtd-mac-address = <&factory 0x28>;
	pinctrl-names = "default";
	pinctrl-0 = <&pa_pins>;
};
```
which boosted reception to -48 dBm (and dramatically increased the iper3 test throughput x4).

By the way, the 5 GHz radio reads -42 dBm for an AP in the room next door, which seems good enough.

Do you prefer to add it to this PR, or shall I open a new one? Up to you :+1:

adrianschmutzler commented:
Do a new one. That's a separate topic, and this one here won't be fast...

adrianschmutzler commented:
@rogerpueyo 
As an interim solution, one could go the same way as suggested in https://github.com/openwrt/openwrt/pull/2353#issuecomment-528346019

I've created a branch for this in
https://github.com/adrianschmutzler/openwrt/commits/dir810bypass
https://github.com/adrianschmutzler/openwrt/commit/acd92ee17202ef81a4f2878af93d443df83063e0

Would be nice if you could test this.

adrianschmutzler commented:
@blocktrron One should think about adding the first patch of this patchset already.
I will fix the wmac MAC address in any case and already provide the correct settings for 5 GHz if the driver is fixed upstream eventually (and the second patch here is not desired).

adrianschmutzler commented:
rebased

adschm committed:
ramips: fix WiFi MAC addresses for D-Link DIR-810L

So far, WiFi MAC addresses for this device have been set up from
caldata. However, this returns values which do not look like MAC
addresses. They also do not match stock firmware:

wlan0 (5.0): 00:11:22:00:17:D0 from 0x8004
wlan1 (2.4): 00:11:22:00:17:CD from 0x4 (and 0x2e)

It looks like the only valid MAC address on this device is at 0x28.

So, this patch changes setup to calculate addresses based on the
value at 0x28:

lan: *:0A (flash, label)
wan: *:0B (flash + 1)
wifi2: *:0A (flash)
wifi5: *:0C (flash + 2)

Thanks to Roger Pueyo Centelles <roger.pueyo@guifi.net> for
investigating this on his devices.

Signed-off-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>

adrianschmutzler commented:
@rogerpueyo I've changed the patches to include the officially merged mt76 patch now.
Can you re-test, please?

rogerpueyo commented:
@adrianschmutzler: perfect!

Current snapshot:
```
root@OpenWrt:~# iw dev
phy#1
	Interface wlan1
		ifindex 4
		wdev 0x100000001
		addr 00:11:22:00:17:cd
		type managed
		txpower 0.00 dBm
phy#0
	Interface wlan0
		ifindex 3
		wdev 0x1
		addr 00:11:22:00:17:d0
		type managed
		txpower 0.00 dBm
		multicast TXQ:
			qsz-byt	qsz-pkt	flows	drops	marks	overlmt	hashcol	tx-bytes	tx-packets
			0	0	0	0	0	0	0	0		0
```

Your branch:
```
root@OpenWrt:~# iw dev
phy#1
	Interface wlan1
		ifindex 4
		wdev 0x100000001
		addr c8:d3:a3:64:ce:0a
		type managed
		txpower 0.00 dBm
phy#0
	Interface wlan0
		ifindex 3
		wdev 0x1
		addr c8:d3:a3:64:ce:0c
		type managed
		txpower 0.00 dBm
		multicast TXQ:
			qsz-byt	qsz-pkt	flows	drops	marks	overlmt	hashcol	tx-bytes	tx-packets
			0	0	0	0	0	0	0	0		0
```

adrianschmutzler commented:
@blocktrron @nbd168 This should include upstream changes properly now (pulling latest mt76) and should be ready for merge.

adrianschmutzler commented:
Recent mt76 update has been merged, so patch 1 can be omitted. Patch 2 still applies.

blocktrron commented:
Merged, thanks!
